vm-allocator/src/gsi.rs

```diff
@@ -67,17 +73,36 @@ impl GsiAllocator {
     pub fn new() -> Self {
         GsiAllocator {
             next_irq: arch::IRQ_BASE,
-            next_gsi: arch::IRQ_BASE,
+            gsi_base: arch::IRQ_BASE,
+            gsi_bitmap: Arc::new(Mutex::new(BitVec::from_elem(KVM_MAX_IRQ_ROUTES, false))),
         }
     }
 
     /// Allocate a GSI
     pub fn allocate_gsi(&mut self) -> Result<u32> {
-        let gsi = self.next_gsi;
-        self.next_gsi = self.next_gsi.checked_add(1).ok_or(Error::Overflow)?;
+        let mut gsi_bitmap = self.gsi_bitmap.lock().unwrap();
+        let mut gsi = GSI_INVALID;
+        for i in self.gsi_base as usize..gsi_bitmap.len() {
+            if !gsi_bitmap[i] {
+                gsi = i as u32;
+                break;
+            }
+        }
+        gsi_bitmap.set(gsi as usize, true);
+
         Ok(gsi)
     }
 
+    /// Free a GSI
+    pub fn free_gsi(&mut self, gsi: u32) {
+        self.gsi_bitmap.lock().unwrap().set(gsi as usize, false);
+    }
+
+    /// Enable gsi bit
+    pub fn enable_gsi_bit(&mut self, gsi: u32) {
```

Comment:
It's a very strange layering issue to have this API here - it would be better to have enable/disable.

Reply:
fixed, thx a lot.